README.md
# Home

<a name=""><img height="36px" src="https://forthebadge.com/images/featured/featured-built-with-love.svg" align="right" alt="Built With 🧡" /></a>

**`somos-py`** is a [Somos sequence](https://en.wikipedia.org/wiki/Somos_sequence) generator for <sub><img height="17px" src="https://api.iconify.design/skill-icons:python-dark.svg" alt="" hspace="2px" /></sub>Python.

If you don't know about **Somos sequences**, they are a series of [mathematical sequences](https://en.wikipedia.org/wiki/Sequence) generated by a [recurrence equation](https://en.wikipedia.org/wiki/Recurrence_relation). Like the well-known [Fibonacci sequence](https://en.wikipedia.org/wiki/Fibonacci_sequence), the next term is defined by some previous terms. You can read [↓ below](#somos-sequences) for more information.

<a name=""><img height="24px" src="https://img.shields.io/pypi/v/somos?style=for-the-badge&label=Version&labelColor=hsla(220%2C40%25%2C50%25%2C15%25)&color=fe7d37" /></a><a href="https://github.com/thatgaypigeon/somos-py/releases/" ><sup><sub><img height="18px" src="https://api.iconify.design/octicon:link-external-16.svg?color=%23ffa724" alt="↗" hspace="6px" /><sup></sub></a>

Check out the [**latest version**](https://github.com/thatgaypigeon/somos-py/releases/latest)!

<a name=""><img height="24px" src="https://img.shields.io/github/license/thatgaypigeon/somos-py?style=for-the-badge&labelColor=hsla(220%2C40%25%2C50%25%2C15%25)&color=FFA724" /></a><a href="https://choosealicense.com/licenses/mit/" ><sup><sub><img height="18px" src="https://api.iconify.design/octicon:link-external-16.svg?color=%23ffa724" alt="↗" hspace="6px" /><sup></sub></a>

This project is released under the **MIT license**. This means you can use, modify, and distribute projects that use this code for free, as long as <u>the original copyright notice and license are included</u>.

You can read more about the MIT license, and what it means if you use this code, [**here**](https://choosealicense.com/licenses/mit/).

<a name=""><img height="24px" src="https://img.shields.io/readthedocs/somos?style=for-the-badge&label=Docs&labelColor=hsla(220%2C40%25%2C50%25%2C15%25)" /></a><a href="https://pigeon.uk.to/somos-py/" ><sup><sub><img height="18px" src="https://api.iconify.design/octicon:link-external-16.svg?color=%23ffa724" alt="↗" hspace="6px" /><sup></sub></a>

You can find documentation on my site [**`pigeon.uk.to/somos-py`**](https://pigeon.uk.to/somos-py) or over at [`somos.readthedocs.io`](https://somos.readthedocs.io)!

## Somos sequences <a href="#top"><img align="right" height="24px" src="https://api.iconify.design/octicon:move-to-top-16.svg?color=%23ffa724" /></a>
> *For more information, see [Wikipedia → Somos sequence](https://en.wikipedia.org/wiki/Somos_sequence).*

**Somos sequences** are a series of mathematical sequences generated by a recurrence equation, like the Fibonacci numbers.

They were first described by Michael Somos in September 1989 as [Problem 1470](https://cms.math.ca/wp-content/uploads/crux-pdfs/Crux_v15n07_Sep.pdf#page=19) in [*Crux Mathematicorum* Volume 15, Issue 7](https://cms.math.ca/wp-content/uploads/crux-pdfs/Crux_v15n07_Sep.pdf) and then later in his paper "[Step into the Elliptic Realm](https://www.cut-the-knot.org/arithmetic/algebra/EllipticSeq.shtml)" in January 2000.

### Definition <a href="#top"><img align="right" height="20px" src="https://api.iconify.design/octicon:move-to-top-16.svg?color=%23ffa724" /></a>
For an integer $k$ larger than 1, the Somos-$k$ sequence is defined by the equation:

$$a_{n}a_{{n-k}}=a_{{n-1}}a_{{n-k+1}}+a_{{n-2}}a_{{n-k+2}}+\cdots +a_{{n-(k-1)/2}}a_{{n-(k+1)/2}}$$

when $k$ is odd; or

$$a_{n}a_{{n-k}}=a_{{n-1}}a_{{n-k+1}}+a_{{n-2}}a_{{n-k+2}}+\cdots +(a_{{n-k/2}})^{2}$$

when $k$ is even, with the initial values $a_{i}=1$ for $i < k$.

### Sequence values <a href="#top"><img align="right" height="20px" src="https://api.iconify.design/octicon:move-to-top-16.svg?color=%23ffa724" /></a>
For $k = 2$ or $3$, the generated sequence is all $1$s.

For $k > 3$, the first $k$ terms are all $1$s.

For $k < 7$, all terms are integers. For $k > 8$, terms will eventually become [fractions](https://en.wikipedia.org/wiki/Fraction). The terms at which Somos-$k$ becomes non-integral are (starting at Somos-$8$): 17, 19, 20, 22, 24, 27, 28, 30, 33, 34, 36, 39, ... (sequence [A030127](https://oeis.org/A030127) in the OEIS).

Below are the first 20 terms of Somos sequences 4 to 10.

<dl>
<dt><b>Somos-4</b> (<a href="https://oeis.org/A006720">A006720</a>)</dt>
<dd>1, 1, 1, 1, 2, 3, 7, 23, 59, 314, 1529, 8209, 83313, 620297, 7869898, 126742987, 1687054711, 47301104551, 1123424582771, 32606721084786, ...</dd>

<dt><b>Somos-5</b> (<a href="https://oeis.org/A006721">A006721</a>)</dt>
<dd>1, 1, 1, 1, 1, 2, 3, 5, 11, 37, 83, 274, 1217, 6161, 22833, 165713, 1249441, 9434290, 68570323, 1013908933, ...</dd>

<dt><b>Somos-6</b> (<a href="https://oeis.org/A006722">A006722</a>)</dt>
<dd>1, 1, 1, 1, 1, 1, 3, 5, 9, 23, 75, 421, 1103, 5047, 41783, 281527, 2534423, 14161887, 232663909, 3988834875, ...</dd>

<dt><b>Somos-7</b> (<a href="https://oeis.org/A006723">A006723</a>)</dt>
<dd>1, 1, 1, 1, 1, 1, 1, 3, 5, 9, 17, 41, 137, 769, 1925, 7203, 34081, 227321, 1737001, 14736001, ...</dd>

<dt><b>Somos-8</b></dt>
<dd>1, 1, 1, 1, 1, 1, 1, 1, 4, 7, 13, 25, 61, 187, 775, 5827, 14815, 420514/7, 28670773/91, 6905822101/2275, ...</dd>

<dt><b>Somos-9</b></dt>
<dd>1, 1, 1, 1, 1, 1, 1, 1, 1, 4, 7, 13, 25, 49, 115, 355, 1483, 11137, 27937, 755098/7, ...</dd>

<dt><b>Somos-10</b></dt>
<dd>1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 5, 9, 17, 33, 65, 149, 413, 1473, 7073, 64785, ...</dd>
</dl>

### In popular culture <a href="#top"><img align="right" height="20px" src="https://api.iconify.design/octicon:move-to-top-16.svg?color=%23ffa724" /></a>
**Somos sequences** were discussed by [Dr Harini Desiraju](https://www.sydney.edu.au/science/about/our-people/academic-staff/harini-desiraju.html), a fellow at the University of Sydney, in a [Numberphile](https://www.youtube.com/@numberphile) video from May 2022.

<center>
<a href="https://www.youtube.com/watch?v=p-HN_ICaCyM" title="The Troublemaker Number - Numberphile on YouTube">
<img src="https://img.youtube.com/vi/p-HN_ICaCyM/0.jpg" alt="The Troublemaker Number - Numberphile" href="https://www.youtube.com/watch?v=p-HN_ICaCyM" />
</a>

<a href="https://www.youtube.com/watch?v=p-HN_ICaCyM">The Troublemaker Number - Numberphile</a> on YouTube
</center>

## Uses <a href="#top"><img align="right" height="24px" src="https://api.iconify.design/octicon:move-to-top-16.svg?color=%23ffa724" /></a>
None, really. This is just a small implementation of a bit of number theory. It's purely recreational. :)

If you *do* happen to find a use for this, please let me know!

## External links <a href="#top"><img align="right" height="20px" src="https://api.iconify.design/octicon:move-to-top-16.svg?color=%23ffa724" /></a>
* [*Crux Mathematicorum* Volume 15, Issue 7](https://cms.math.ca/wp-content/uploads/crux-pdfs/Crux_v15n07_Sep.pdf#page=19) at [cms.math.ca](https://cms.math.ca)
* [Somos sequences](https://en.wikipedia.org/wiki/Somos_sequence) at Wikipedia
* [Somos sequences](https://mathworld.wolfram.com/SomosSequence.html) at MathWorld
* [The Somos Sequence Site](https://faculty.uml.edu//jpropp/somos.html) at [uml.edu](https://uml.edu), by *James Propp*
* [The Troublemaker Number - Numberphile](https://www.youtube.com/watch?v=p-HN_ICaCyM) on YouTube


CHANGELOG.md
# Changelog

All notable changes to this project will be documented here. For more details about each release, see [the documentation](https://somos.readthedocs.io/en/stable/releases/index.html).

Versioning conforms to `MAJOR.MINOR.PATCH[{a|b|rc}N][.devN]`. See also [thatgaypigeon → Versioning](https://github.com/thatgaypigeon/standards/wiki/Versioning).

## Version [`0.1.0`](https://github.com/thatgaypigeon/somos-py/releases/tags/0.1.0) (Nov 2023)
> Full release details [here](https://somos.readthedocs.io/en/stable/releases/0.1.0.html).

<big>**Initial** release of `somos`.</big>

- Added class `Somos`
- Added properties `k`, `sequence`, `trivial`
- Added method `generate_terms()`


tests/test_somos.py
"""Somos tests"""

import unittest

from somos import Sequence


class TestSequence(unittest.TestCase):
    def test_sequence_generation(self):
        seq = Sequence(4, num_terms=10)
        expected_sequence = [1, 1, 1, 1, 2, 3, 7, 23, 59, 314]
        self.assertEqual(seq.sequence, expected_sequence)

    def test_sequence_with_k_2(self):
        seq = Sequence(2, num_terms=10)
        expected_sequence = [1, 1, 1, 1, 1, 1, 1, 1, 1, 1]
        self.assertEqual(seq.sequence, expected_sequence)

    def test_sequence_with_large_k(self):
        seq = Sequence(10, num_terms=10)
        # Check that the sequence starts with 10 ones
        self.assertEqual(seq.sequence[:10], [1] * 10)

    def test_invalid_k(self):
        with self.assertRaises(ValueError):
            Sequence(1)

    def text_negative_k(self):
        with self.assertRaises(ValueError):
            Sequence(-1)

    def test_k_not_int(self):
        with self.assertRaises(TypeError):
            Sequence("4")

    def test_invalid_num_terms(self):
        with self.assertRaises(ValueError):
            Sequence(4, num_terms=0)

    def test_negative_num_terms(self):
        with self.assertRaises(ValueError):
            Sequence(4, num_terms=-1)

    def test_num_terms_not_int(self):
        with self.assertRaises(TypeError):
            Sequence(4, num_terms="10")

    def test_invalid_num_non_triv_terms(self):
        with self.assertRaises(ValueError):
            Sequence(4, num_non_triv_terms=0)

    def test_negative_num_non_triv_terms(self):
        with self.assertRaises(ValueError):
            Sequence(4, num_non_triv_terms=-1)

    def test_num_non_triv_terms_not_int(self):
        with self.assertRaises(TypeError):
            Sequence(4, num_non_triv_terms="10")


if __name__ == "__main__":
    unittest.main()


somos/somos.py
import logging
import math
from fractions import Fraction
from typing import Literal, NamedTuple


class VersionInfo(NamedTuple):
    major: int
    minor: int
    micro: int
    level: Literal["a", "b", "rc"] | None = None
    pre_release: int | None = None
    dev: int | None = None


__version_info__: VersionInfo = VersionInfo(major=0, minor=1, micro=0)
"""Version information."""

__title__ = "somos"
"""Project title."""
__author__ = "Max Kerr (thatgaypigeon)"  # Pigeon43
"""Project author."""
__license__ = "MIT"
"""Project license."""
__copyright__ = "Copyright 2023-present Max Kerr (thatgaypigeon)"  # Pigeon43
"""Project copyright information."""
__version__ = f"{__version_info__.major}.{__version_info__.minor}.{__version_info__.micro}{f'{__version_info__.level}{__version_info__.pre_release}' if __version_info__.level is not None and __version_info__.pre_release is not None else ''}{f'.dev{__version_info__.dev}' if __version_info__.dev is not None else ''}"
"""Project version."""

logging.getLogger(__name__).addHandler(logging.NullHandler())

del logging, NamedTuple, Literal, VersionInfo

__all__ = ["Sequence"]


class Sequence:
    """Represents a somos sequence object.

    .. container:: operations

        .. describe:: x == y

            Checks if two sequences are equal.

        .. describe:: x != y

            Checks if two sequences are not equal.

        .. describe:: str(x)

            Returns the sequence list, comma separated.

        .. describe:: repr(x)

            Returns the sequence's string representation.

    A number of options can be passed to the :class:`Sequence`.

    Parameters
    ------------
    k: :class:`int`
        The integer ``k`` for which to produce the sequence ``somos-k``.
    num_terms Optional[:class:`int`] (default: :paramref:`k` ``+`` :paramref:`num_non_triv_terms`)
        The number of terms to generate.
    num_non_triv_terms: Optional[:class:`int`] (default: ``10``)
        The number of non-trivial terms to generate.
        This excludes the first :paramref:`k` terms since ``a(n) for n < k`` always returns ``1``,
        but does not include all terms for the trivial sequences ``somos-2`` and ``somos-3``.

    Attributes
    ------------
    k: :class:`int`
        The integer ``k`` for which to produce the sequence ``somos-k``.
    sequence: List[:class:`int | Fraction`]
        The list of numbers within the sequence.
    trivial: :class:`bool`
        Whether or not the sequence is trivial. ``True`` for ``k = 2`` and ``k = 3``, else ``False``.
    """

    __slots__: tuple = ("_k", "_sequence", "_trivial", "_ctx")

    def __init__(self, k: int, num_terms: int | None = None, num_non_triv_terms: int | None = None) -> None:
        if not isinstance(k, int):
            # Catch k is not int
            raise TypeError(f"Expected int, received {k.__class__.__name__} instead.")
        elif not isinstance(num_terms, int) and num_terms is not None:
            # Catch num_terms is not int
            raise TypeError(
                f"Expected int, received {num_terms.__class__.__name__} instead."
            )
        elif not isinstance(num_non_triv_terms, int) and num_non_triv_terms is not None:
            # Catch num_non_triv_terms is not int
            raise TypeError(
                f"Expected int, received {num_non_triv_terms.__class__.__name__} instead."
            )

        if k <= 1:
            # Catch values of k below 2
            raise ValueError(
                f"A k value of {k} is invalid. Please use an integer greater than 1."
            )
        elif num_terms is not None and num_terms < 1:
            # Catch values of num_terms below 1
            raise ValueError(
                f"A num_terms value of {num_terms} is invalid. Please use an integer greater than 0."
            )
        elif num_non_triv_terms is not None and num_non_triv_terms < 1:
            # Catch values of num_non_triv_terms below 1
            raise ValueError(
                f"A num_non_triv_terms value of {num_non_triv_terms} is invalid. Please use an integer greater than 0."
            )

        self._k: int = k
        self._trivial: bool = True if self.k < 3 else False
        self._sequence: list[int | Fraction] = []

        self.generate_terms(
            num_terms
            if num_terms
            else (self.k + (num_non_triv_terms if num_non_triv_terms else 10))
        )

    def _term_n(self, n: int) -> int | Fraction:
        """Returns term ``n`` of sequence ``somos-k``.

        .. warning::

            This is an internal method, not intended to be accessed by the user.

            To get term ``n`` of sequence ``somos-k``, use ``Sequence.sequence[n]``.
            Bear in mind ``Sequence.sequence`` is zero-indexed.
        """
        a: list[int | Fraction] = self.sequence
        k: int = self.k

        if n < k:
            # Catch trivial cases. (The first ``k`` terms of sequence ``somos-k`` are trivial ``1``s.)
            return 1

        _range: range = range(1, math.floor(k / 2) + 1)
        _prev_terms: list[int | Fraction] = [a[n - j] * a[n - (k - j)] for j in _range]
        a_n: int | Fraction = Fraction(sum(_prev_terms), a[n - k])
        return a_n

    def __eq__(self, other: object) -> bool:
        return isinstance(other, Sequence) and self.k == other.k

    def __ne__(self, other: object) -> bool:
        return not self.__eq__(other)

    def __str__(self) -> str:
        return ", ".join([str(n) for n in self.sequence])

    def __repr__(self) -> str:
        return f"<somos.Sequence k={self.k} sequence={self.sequence} trivial={self.trivial}>"

    @property
    def k(self) -> int:
        """The integer ``k`` for which to produce the sequence ``somos-k``."""
        return self._k

    @property
    def sequence(self) -> list[int | Fraction]:
        """The list of numbers within the sequence."""
        return self._sequence

    @property
    def trivial(self) -> bool:
        """If the sequence is trivial. ``True`` for ``k = 2`` and ``k = 3``, else ``False``."""
        return self._trivial

    def generate_terms(self, count: int) -> None:
        """Continues generating the ``sequence`` for the specified number of terms.

        Arguments
        ---------
        count: :class:`int`
            The number of terms to generate.

        Raises
        ------
        ``ValueError``: "A count value of 'count' is invalid. Please use an integer greater than 0."
            For ``count < 1``. (``count is not int`` already caught.)

        Examples
        --------
        Generate base sequence with ``3`` terms. (Total 3 terms)
        ```
        >>> seq = Sequence(5, num_terms=3)
        # 1, 1, 1
        ```

        Generate ``5`` more terms. (Total 8 terms)
        ```
        >>> seq.generate_terms(5)
        # 1, 1, 1, 1, 1, 2, 3, 5
        ```

        Generate ``10`` more terms. (Total 18 terms)
        ```
        >>> seq.generate_terms(10)
        # 1, 1, 1, 1, 1, 2, 3, 5, 11, 37, 83, 274, 1217, 6161, 22833, 165713, 1249441, 9434290
        ```"""
        if not isinstance(count, int):
            # Catch count is not int
            raise TypeError(
                f"Expected int, received {count.__class__.__name__} instead."
            )
        elif count < 1:
            # Catch values of count below 1
            raise ValueError(
                f"A count value of {count} is invalid. Please use an integer greater than 0."
            )
        elif self.k <= 3:
            # Catch trivial cases. (All terms of sequences somos-2 and somos-3 are trivial 1s.)
            self.sequence.extend([1 for i in range(count)])
            return

        for n in range(len(self.sequence), len(self.sequence) + count):
            self.sequence.append(self._term_n(n))
